src/key_info_managers/sqlite_manager/mod.rs

```diff
@@ -0,0 +1,597 @@
+                key_info              BLOB NOT NULL,
+                key_info_version      INTEGER NOT NULL,
+                PRIMARY KEY (authenticator_id, application_name, key_name)
+            )
+            ",
+            [],
+        )?;
+
+        let mut stmt = conn.prepare("SELECT * FROM key_mapping")?;
+        let key_iter = stmt.query_map([], |row| {
+            let key_identity = KeyIdentity::new(
+                ApplicationIdentity::new(
+                    row.get("application_name")?,
+                    i64_to_auth_type(row.get("authenticator_id")?).map_err(|e| {
+                        format_error!("Failed to get AuthType from authenticator_id.", e);
+                        let error = Box::new(Error::new(ErrorKind::InvalidData, e));
+                        RusqliteError::FromSqlConversionFailure(64, Integer, error)
+                    })?,
+                ),
+                ProviderIdentity::new(row.get("provider_uuid")?, row.get("provider_name")?),
+                row.get("key_name")?,
+            );
+            let key_info_blob: Vec<u8> = row.get("key_info")?;
+
+            // TODO: Change this to (protobuf?) version once format has been decided.
+            let key_info = bincode::deserialize(&key_info_blob[..]).map_err(|e| {
+                format_error!("Error deserializing key info", e);
+                RusqliteError::FromSqlConversionFailure(key_info_blob.len(), Blob, e)
+            })?;
+
+            Ok(KeyInfoRecord {
+                key_identity,
+                key_info,
+            })
+        })?;
+
+        // Add keys to key_store cache
+        for key_info_record in key_iter {
+            let key_info_record = key_info_record?;
+            let _ = key_store.insert(key_info_record.key_identity, key_info_record.key_info);
+        }
+
+        if !crate::utils::GlobalConfig::log_error_details() {
+            info!(
+                "SQLiteKeyInfoManager - Found {} key info mapping records",
+                key_store.len()
+            );
+        }
+
+        Ok(SQLiteKeyInfoManager {
+            key_store,
+            database_path,
+        })
+    }
+
+    /// Saves the KeyIdentity and KeyInfo to the database.
+    /// Inserts a new record to the database `key_mapping` table.
+    fn save_mapping(
+        &self,
+        key_identity: &KeyIdentity,
+        key_info: &KeyInfo,
+    ) -> rusqlite::Result<(), RusqliteError> {
+        let conn = Connection::open(&self.database_path)?;
+
+        // TODO: Change this to (protobuf?) version once format has been decided.
+        let key_info_blob = bincode::serialize(&key_info).map_err(|e| {
+            format_error!("Error serializing key info", e);
+            RusqliteError::ToSqlConversionFailure(e)
+        })?;
+
+        // Insert the new key mapping, if a record does not exist.
+        // If one does exist, update the existing record.
+        let _ = conn.execute(
+            "
+            INSERT OR IGNORE INTO
```

Comment:
Does the `INSERT OR IGNORE` mean that if one of the insertions fails, it fails silently (i.e., no error is returned by the KIM)?

Reply:
The `INSERT OR IGNORE` clause will attempt to `INSERT` first; if there is already a record that exists then this will fail (silently, not throwing a `RusqliteError`). For every request it attempts to `UPDATE`. If the `UPDATE` fails, this will not be silent and will generate a `RusqliteError`.

Essentially it's "making sure the record exists", and then "updating it to the correct value".

Reply:
Thanks for the explanation! 

This might sound strange, but would it make sense to split this into the two operations, `INSERT` and `UPDATE`, let the `INSERT` fail and "manually" ignore it if it does (by, for example, dropping the result from `.execute()`)? Don't know if this makes the whole thing less efficient or if it's just good practice for SQL queries to be done together like this, though

Reply:
On further investigation there is a more semantically correct `REPLACE` operation.
https://dev.mysql.com/doc/refman/8.0/en/replace.html
This should result in one query such as something like:
```
REPLACE INTO
    `key_mapping`
    (`authenticator_id`, `application_name`, `provider_uuid`, `provider_name`, `key_name`, `key_info`, `key_info_version`)
VALUES
    (?1, ?2, ?3, ?4, ?5, ?6, ?7);
```
If the record does not exist, it inserts it. If it does, then it deletes the existing record and inserts the new one. (This occurs by matching on the primary/unique indexes which in our case are `authenticator_id`-`application_name`-`key_name`)

Reply:
> I'll have a look into implementing this and whether it is supported in SQLite.

Update: seems to work fine. Will push a revised PR commit.